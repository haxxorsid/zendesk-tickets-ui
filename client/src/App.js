import { useEffect, useState } from 'react';
import './styles/App.css';
import Tickets from './components/Tickets.js';
import TicketDetails from './components/TicketDetails';
import loader from './assets/loader.svg';
import { getTickets } from './common/api';
import { paginateTickets } from './common/helper';

function App() {

  // state variable getters and setters
  const [tickets, setTickets] = useState(null)
  const [paginatedTickets, setPaginatedTickets] = useState(null);
  const [error, setError] = useState(null);
  const [currentPage, setCurrentPage] = useState(0);
  const [countPerPage] = useState(25);
  const [selectedTicket, setSelectedTicket] = useState(undefined);
  const [numberOfPages, setNumberOfPages] = useState([1]);

  // get tickets, set tickets once if success, otherwise set error
  useEffect(() => {
    getTickets()
      .then(res => {
        if (typeof res === "string") {
          setError(res);
        } else {
          setTickets(res)
        }
      })
      .catch(error => {
        setError(error.message)
      })
  }, [])

  // once ticket are fetched, group tickets in arrays depending on count per page
  useEffect(() => {
    tickets && setPaginatedTickets(paginateTickets(tickets, countPerPage))
  }, [tickets, countPerPage])

  // calculate number of pages for pagination once tickets have been grouped
  useEffect(() => {
    paginatedTickets && setNumberOfPages(paginatedTickets.map((page, index) => index))
  }, [paginatedTickets])

  function onPageChange(number) {
    setCurrentPage(number)
  }

  // set selected ticket, this selectedTicket will be passed on to the modal to view ticket details
  function onSelectTicket(id) {
    setSelectedTicket(tickets.filter(ticket => ticket.id === id))
  }

  // clear ticket if view ticket details modal was closed
  function onSelectedTicketClear(id) {
    setSelectedTicket(undefined)
  }

  return (
    <div className="App">
      <header className="App-header">
        <h3 className="header-title">Zendesk Tickets UI</h3>
      </header>
      {error ? (
        <p className="error">
          {error}
        </p>
      ) : (
        <div>
          {tickets ? (
            <Tickets
              tickets={paginatedTickets ? paginatedTickets[currentPage] : null}
              count={tickets && tickets.length}
              countPerPage={paginatedTickets ? paginatedTickets[currentPage].length : null}
              onSelectTicket={onSelectTicket}
              currentPage={currentPage}
              numberOfPages={numberOfPages}
              onPageChange={onPageChange}
            />
          ) : (
            <img className="loader" src={loader} alt="Loading..."/>
          )}
          <TicketDetails selectedTicket={selectedTicket} onSelectedTicketClear={onSelectedTicketClear} />
        </div>
      )
      }
    </div>
  );
}

export default App;
