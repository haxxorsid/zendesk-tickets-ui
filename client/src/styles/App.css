.App {
  text-align: center;
  background-size: cover;
  height: 100vh;
}


.tickets {
  color: #03363D;
  background: #fff;
  margin: 20px auto;
  padding: 10px;
  width: 600px;
}

.ticket {
  border: 1px solid lightgray;
  cursor: pointer;
  margin: 5px;
}

.ticket-header {
  text-align: left;
  padding-left: 10px;
}

.date {
  color: grey;
}

.error {
  background: #fff;
  width: 400px;
  padding: 5px;
  margin: 40px auto;
  color: red;
}


.paginator {
  max-width: 500px;
  list-style: none;
  display: flex;
  justify-content: center;
  margin: 1rem auto;
  padding: 5px;
}

.paginator-item {
  border: .5px solid lightgray;
  padding: 10px;
  user-select: none;
  cursor: pointer;
}

.current-page {
  background: #03363D;
  color: #fff;
}

.modal {
  background: #fff;
  border: .5px solid lightgray;
  outline: none;
  padding: 10px;
  text-align: center;
  width: 800px;
}

.button {
  background: #03363D;
  width: 100px;
  color: #fff;
  padding: 10px;
  font-size: 15px;
  margin: 0 auto;
}

.ticket-details-description {
  padding: 15px;
  text-align: left;
  margin: 10px;
}

.loader {
  margin: 200px auto;
}

.right {
  float: right;
  padding-right: 10px;
}

.id {
  padding-right: 5px;
  color: #fff;
  padding-left: 5px;
  background: #03363D;
}

.status {
  padding-right: 5px;
  color: #fff;
  padding-left: 5px;
  background: #2a6871;
  margin-left: 5px;
  margin-right: 5px;
}

.App-header {
  background:  #03363D;
  color: #ffffff;
  display: flex;
  padding: 5px;
}

.header-title {
  font-size: 1.5em;
  margin: 10px auto;
  padding-top: 5px;
  width: 500px;
}

.ReactModalPortal > div {
  opacity: 0;
}

.ReactModalPortal .ReactModal__Overlay {
  align-items: center;
  display: flex;
  justify-content: center;
}

.ReactModalPortal .ReactModal__Overlay--before-close {
  opacity: 0;
}

.ReactModalPortal .ReactModal__Overlay--after-open {
  opacity: 1;
}