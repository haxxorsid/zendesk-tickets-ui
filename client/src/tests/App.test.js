import React from 'react';
import renderer from 'react-test-renderer'
import App from '../App';

it('Should render App correctly', () => {
  const tree = renderer
    .create(<App />)
    .toJSON();
  expect(tree).toMatchSnapshot()
})