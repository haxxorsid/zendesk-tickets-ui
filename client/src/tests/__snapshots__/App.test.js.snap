// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render App correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h3
      className="header-title"
    >
      Zendesk Tickets UI
    </h3>
  </header>
  <div>
    <img
      alt="Loading..."
      className="loader"
      src="loader.svg"
    />
  </div>
</div>
`;
